### PR TITLE
Allocate the escape table lazily

### PR DESCRIPTION
**Pull request by andralex**
According to [heaptrack](http://manpages.ubuntu.com/manpages/bionic/man1/heaptrack.1.html), the `EscapeTable` allocations count for hundreds of megabytes in large builds. This reduces such a large build's peak memory consumption from 22.18 GB to 21.70 GB.

### CODE REVIEW

File: src/dmd/dmodule.d [hunk 1]
```diff
@@ -1565,6 +1564,16 @@ extern (C++) final class Module : Package
...
         }
     }
+
+    /** Lazily initializes and returns the escape table.
+    Turns out it eats a lot of memory.
+    */
+    Escape* escapetable()
+    {
+        if (!_escapetable)
+            _escapetable = new Escape;
```
Comment (UplinkCoder) [inline src/dmd/dmodule.d:1574]: Parens.

**Comment by UplinkCoder** [comment]
this is an abi breaking change.

**Comment by wilzbach** [comment]
> this is an abi breaking change. 

Why do you think that?

**Comment by UplinkCoder** [comment]
abi of dmd.
not of executables generated by dmd.

File: src/dmd/dmodule.d (hunk 1 above)
Comment (andralex) [reply src/dmd/dmodule.d:1574]: Braces? Looking through the same module, it consistently does not brace single statements (sometimes annoyingly, as in if/else, one with braces and one without). But, when in Rome.

**Comment by andralex** [comment]
@ibuclaw @kinke ABI issues to worry about?

File: src/dmd/dmodule.d [hunk 2]
```diff
@@ -1565,6 +1564,16 @@ extern (C++) final class Module : Package
             buf.prependstring(package_.ident.toChars());
         }
     }
+
+    /** Lazily initializes and returns the escape table.
+    Turns out it eats a lot of memory.
+    */
+    Escape* escapetable()
```
Comment (ibuclaw) [inline src/dmd/dmodule.d:1571]: Just mark this as `extern (D)` and we're all good.

Comment (andralex) [reply src/dmd/dmodule.d]: done, thanks

**Comment by RazvanN7** [comment]
@andralex To fix the failing circleCI test you need to update `src/dmd/frontend.h` to reflect the name change `escapeTable -> _escapeTable` [here](https://github.com/dlang/dmd/blob/master/src/dmd/frontend.h#L2890).

**Review by RazvanN7 — APPROVED** [review]

**Comment by andralex** [comment]
@RazvanN7 done, thanks!

**Comment by UplinkCoder** [comment]
There a multi-threading concerns with this.
Please isolate a place where you can do early initialisation.
(i.e. initialisation before use as opposed to initialisation during first use).

As far as I am aware this is only used during ddoc generation?
if so initialize when you know that ddoc is being generated.

**Comment by ibuclaw** [comment]
Why would threads be a concern? The compiler is single threaded, and ddoc generation seems to be one place that you wouldn't need parallelism anyway.

File: src/dmd/dmodule.d (hunk 1 above)
Comment (andralex) [reply src/dmd/dmodule.d]: OK. When in Rome.

File: src/dmd/backend/outbuf.d [hunk 3]
```diff
@@ -43,7 +45,10 @@ struct Outbuffer
     @trusted
     void dtor()
     {
-        free(buf);
+        if (length >= huge)
+            free(buf);
+        else
+            free(buf);
```
Comment (RazvanN7) [inline src/dmd/backend/outbuf.d:51]: What is the point of this condition if `free(buf)` is called unconditionally? Also, is this related in any way?

Comment (andralex) [reply src/dmd/backend/outbuf.d:51]: ouch my work in progress spilled... sorry! fixing now

**Review by RazvanN7 — APPROVED** [review]

File: src/dmd/dmodule.d [hunk 4]
```diff
@@ -545,7 +545,7 @@ extern (C++) final class Module : Package
     Identifiers* versionidsNot; // forward referenced version identifiers
 
     MacroTable macrotable;      // document comment macros
-    Escape* escapetable;        // document comment escapes
+    Escape* _escapetable;       // document comment escapes
```
Comment (TungstenHeart) [inline src/dmd/dmodule.d:548]: although the chance that someone one day touches this directly are low it should be set `private ` if possible.

File: src/dmd/dmodule.d [hunk 5]
```diff
@@ -1565,6 +1564,16 @@ extern (C++) final class Module : Package
             buf.prependstring(package_.ident.toChars());
         }
     }
+
+    /** Lazily initializes and returns the escape table.
+    Turns out it eats a lot of memory.
+    */
+    extern(D) Escape* escapetable()
```
Comment (WalterBright) [inline src/dmd/dmodule.d:1571]: Curiously, this same code appears in `parseEscapes()` in doc.d. I also wonder why the memory allocated for that needs to be retained and is not recycled.
